fix NED ENU conversion

Stapelzeiger opened the pull request:
The NED to ENU conversion flips the axes by 180 degrees around an axis that is between the x & y axes.

This was correctly implemented for the angular rate & acceleration, but not the quaternion of the estimated attitude.

This PR fixes the quaternion transformation.

Stapelzeiger committed:
fix NED ENU conversion

Stapelzeiger commented:
likely fixes https://github.com/dawonn/vectornav/issues/137

dawonn commented:
Would it be possible to implement this without the additional eigen dependentcy?

Stapelzeiger commented:
It should be very doable to write out the expression by hand. I won't have time to do it in the next weeks though.

Stapelzeiger committed:
use tf2 instead of Eigen

Stapelzeiger committed:
fix frame for other quantities in body axes

Stapelzeiger commented:
I've changed it to use tf2's quaternion math.
I've also updated other places that publish quantities in the body frame (velocity & magnetic field) to be consistent, since they share the same frame_id.

Stapelzeiger commented on vectornav/src/vn_sensor_msgs.cc:
```diff
@@ -309,20 +303,21 @@ void VnSensorMsgs::sub_vn_common(const vectornav_msgs::msg::CommonGroup::SharedP
-      msg.pose.pose.orientation = toMsg(q_enu2ecef * q);
+      tf2::Quaternion q_frd2ecef =q_ned2ecef * q_frd2ned;
+      msg.pose.pose.orientation = tf2::toMsg(q_frd2ecef);
     }
```
This enu2ecef rotation was not correct from what I can tell. This change should produce the correct Right-Front-Up to ECEF (in end mode) or Forward-Right-Down (in NED mode).

Stapelzeiger commented:
I've checked quaternion, angular rate, and acceleration of both IMU messages; the quaternion of the PoseWithCovariance message; and the magnetic field on a VN100 in both ENU & NED mode. The outputs look correct to me.

noahdhan commented on vectornav/src/vn_sensor_msgs.cc:
```diff
@@ -187,16 +179,13 @@ void VnSensorMsgs::sub_vn_common(const vectornav_msgs::msg::CommonGroup::SharedP
-      fromMsg(msg_in->quaternion, q);
-      q_ned2enu.setRPY(M_PI, 0.0, -M_PI / 2);
-      msg.orientation = toMsg(q_ned2enu * q);
+      msg.orientation = msg_in->quaternion;
```
Thanks for making this PR! I also came to the conclusion that the IMU quaternion message is wrong for NED and was going to make this same suggestion

dawonn commented:
Thanks for confirmation!
